bug: ibis can't handle response from mssql when statement is `CAST('2024-01-01 00:00:00' AS DATETIMEOFFSET)`

### What happened?

Ibis will raise an error when I use a statement with `CAST AS DATETIMEOFFSET` for `mssql`.

```python
import ibis

ibis.mssql.connect(...).sql("SELECT CAST('2024-01-01 00:00:00' AS DATETIMEOFFSET)").to_pandas()
```

### What version of ibis are you using?

9.2.0

### What backend(s) are you using, if any?

mssql

### Relevant log output

```sh
return self.connection.sql(sql).limit(limit).to_pandas()
../.venv/lib/python3.11/site-packages/ibis/backends/sql/__init__.py:203: in sql
    schema = self._get_schema_using_query(query)
../.venv/lib/python3.11/site-packages/ibis/backends/mssql/__init__.py:247: in _get_schema_using_query
    return sch.Schema(schema)
../.venv/lib/python3.11/site-packages/ibis/common/bases.py:72: in __call__
    return cls.__create__(*args, **kwargs)
../.venv/lib/python3.11/site-packages/ibis/common/grounds.py:119: in __create__
    kwargs = cls.__signature__.validate(cls, args, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Signature (fields: FrozenOrderedDict[str, DataType])>
func = <class 'ibis.expr.schema.Schema'>
args = ({None: Timestamp(timezone='UTC', scale=7, nullable=True)},), kwargs = {}

    def validate(self, func, args, kwargs):
        """Validate the arguments against the signature.
    
        Parameters
        ----------
        func : Callable
            Callable to validate the arguments for.
        args : tuple
            Positional arguments.
        kwargs : dict
            Keyword arguments.
    
        Returns
        -------
        validated : dict
            Dictionary of validated arguments.
    
        """
        try:
            bound = self.bind(*args, **kwargs)
            bound.apply_defaults()
        except TypeError as err:
            raise SignatureValidationError(
                "{call} {cause}\n\nExpected signature: {sig}",
                sig=self,
                func=func,
                args=args,
                kwargs=kwargs,
            ) from err
    
        this, errors = {}, []
        for name, value in bound.arguments.items():
            param = self.parameters[name]
            pattern = param.annotation.pattern
    
            result = pattern.match(value, this)
            if result is NoMatch:
                errors.append((name, value, pattern))
            else:
                this[name] = result
    
        if errors:
>           raise SignatureValidationError(
                "{call} has failed due to the following errors:{errors}\n\nExpected signature: {sig}",
                sig=self,
                func=func,
                args=args,
                kwargs=kwargs,
                errors=errors,
            )
E           ibis.common.annotations.SignatureValidationError: Schema({None: Timestamp(timezone='UTC', scale=7, nullable=True)}) has failed due to the following errors:
E             `fields`: {None: Timestamp(timezone='UTC', scale=7, nullable=True)} is not matching GenericMappingOf(key=InstanceOf(type=<class 'str'>), value=CoercedTo(type=<class 'ibis.expr.datatypes.core.DataType'>, func=<bound method DataType.__coerce__ of <class 'ibis.expr.datatypes.core.DataType'>>), type=CoercedTo(type=<class 'ibis.common.collections.FrozenOrderedDict'>, func=<class 'ibis.common.collections.FrozenOrderedDict'>))
E           
E           Expected signature: Schema(fields: FrozenOrderedDict[str, DataType])

../.venv/lib/python3.11/site-packages/ibis/common/annotations.py:497: SignatureValidationError
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/ibis/backends/mssql/__init__.py
+++ b/ibis/backends/mssql/__init__.py
@@ -307,6 +307,9 @@ class Backend(SQLBackend, CanCreateCatalog, CanCreateDatabase, CanCreateSchema):
             elif newtyp.is_timestamp():
                 newtyp = newtyp.copy(scale=scale)
 
+            if name is None:
+                name = util.gen_name("col")
+
             schema[name] = newtyp
 
         return sch.Schema(schema)

--- a/ibis/backends/mssql/tests/test_client.py
+++ b/ibis/backends/mssql/tests/test_client.py
@@ -241,3 +241,27 @@ def test_from_url():
     )
     result = new_con.sql("SELECT 1 AS [a]").to_pandas().a.iat[0]
     assert result == 1
+
+
+def test_dot_sql_with_unnamed_columns(con):
+    expr = con.sql(
+        "SELECT CAST('2024-01-01 00:00:00' AS DATETIMEOFFSET), 'a' + 'b', 1 AS [col42]"
+    )
+
+    schema = expr.schema()
+    names = schema.names
+
+    assert len(names) == 3
+
+    assert names[0].startswith("ibis_col")
+    assert names[1].startswith("ibis_col")
+    assert names[2] == "col42"
+
+    assert schema.types == (
+        dt.Timestamp(timezone="UTC", scale=7),
+        dt.String(nullable=False),
+        dt.Int32(nullable=False),
+    )
+
+    df = expr.execute()
+    assert len(df) == 1

--- a/ibis/backends/tests/test_dot_sql.py
+++ b/ibis/backends/tests/test_dot_sql.py
@@ -13,7 +13,11 @@ import ibis.common.exceptions as com
 from ibis import _
 from ibis.backends import _get_backend_names
 from ibis.backends.tests.base import PYTHON_SHORT_VERSION
-from ibis.backends.tests.errors import GoogleBadRequest, OracleDatabaseError
+from ibis.backends.tests.errors import (
+    ExaQueryError,
+    GoogleBadRequest,
+    OracleDatabaseError,
+)
 
 pd = pytest.importorskip("pandas")
 tm = pytest.importorskip("pandas.testing")
@@ -333,6 +337,10 @@ def test_embedded_cte(alltypes, ftname_raw):
 
 
 @dot_sql_never
+@pytest.mark.never(["exasol"], raises=ExaQueryError, reason="backend requires aliasing")
+@pytest.mark.never(
+    ["oracle"], raises=OracleDatabaseError, reason="backend requires aliasing"
+)
 def test_unnamed_columns(con):
     sql = "SELECT 'a', 1 AS col42"
     sgexpr = sg.parse_one(sql, read="duckdb")

--- a/ibis/backends/tests/test_dot_sql.py
+++ b/ibis/backends/tests/test_dot_sql.py
@@ -330,3 +330,22 @@ def test_embedded_cte(alltypes, ftname_raw):
     expr = alltypes.sql(sql, dialect="duckdb")
     result = expr.head(1).execute()
     assert len(result) == 1
+
+
+@dot_sql_never
+def test_unnamed_columns(con):
+    sql = "SELECT 'a' || 'b', 1 AS col42"
+    sgexpr = sg.parse_one(sql, read="duckdb")
+    expr = con.sql(sgexpr.sql(con.dialect))
+
+    schema = expr.schema()
+    names = schema.names
+    types = schema.types
+
+    assert len(names) == 2
+
+    assert names[0]
+    assert names[1] == "col42"
+
+    assert types[0].is_string()
+    assert types[1].is_integer()

--- a/ibis/backends/tests/test_dot_sql.py
+++ b/ibis/backends/tests/test_dot_sql.py
@@ -334,7 +334,7 @@ def test_embedded_cte(alltypes, ftname_raw):
 
 @dot_sql_never
 def test_unnamed_columns(con):
-    sql = "SELECT 'a' || 'b', 1 AS col42"
+    sql = "SELECT 'a', 1 AS col42"
     sgexpr = sg.parse_one(sql, read="duckdb")
     expr = con.sql(sgexpr.sql(con.dialect))
 

